services/BBB.go

```diff
@@ -138,37 +140,73 @@ func (bbb BBBService) ShouldScale(server s.Server) (s.ScaleResource, error) {
 		return targetResource, err
 	}
 
-	// Scaling cores
-	cpuMaxUsageDelta := server.ServerCpuUsage - bbb.Config.Resources.Cpu.MaxUsage
+	// Scaling rules
 
-	if server.ServerCpu >= int32(bbb.Config.Resources.Cpu.MinCores) && cpuMaxUsageDelta <= 0 {
-		targetResource.Cpu.Direction = s.ScaleNone
+	// Rule 1: Scale up cpu if current cpu is below configured minimum for the service
+	if server.ServerCpu < int32(bbb.Config.Resources.Cpu.MinCores) {
+		targetResource.Cpu.Direction = s.ScaleUp
+		targetResource.Cpu.Reason = targetResource.Cpu.Reason + ",Rule 1"
+		targetResource.Cpu.Amount = int32(bbb.Config.Resources.Cpu.MinCores)
 	}
-	if server.ServerCpu < int32(bbb.Config.Resources.Cpu.MinCores) || cpuMaxUsageDelta > 0 {
+
+	// Rule 2: Scale up cpu if current cpu usage is above configured maximum usage for the service
+	// Scale up to either reach cpu usage below the configured maximum usage or to the configured maximum cpu
+	if cpuMaxUsageDelta := server.ServerCpuUsage - bbb.Config.Resources.Cpu.MaxUsage; cpuMaxUsageDelta > 0 {
 		targetResource.Cpu.Direction = s.ScaleUp
+		targetResource.Cpu.Reason = targetResource.Cpu.Reason + ",Rule 2"
 		cpuInc := cpuMaxUsageDelta * float32(server.ServerCpu) / server.ServerCpuUsage
-		targetResource.Cpu.Amount = server.ServerCpu + int32(math.Ceil(float64(cpuInc)))
+		targetHeuristic := server.ServerCpu + int32(math.Ceil(float64(cpuInc)))
+		targetResource.Cpu.Amount = int32(math.Min(float64(targetHeuristic), float64((bbb.Config.Resources.Cpu.MaxCores))))
 	}
 
-	if server.ServerCpu > int32(bbb.Config.Resources.Cpu.MinCores) && participantsCount == 0 {
+	// Rule 3: RuleScale down cpu if current cpu usage is below configured minimum usage for the service
+	if cpuMinUsageDelta := server.ServerCpuUsage - bbb.Config.Resources.Cpu.MinUsage; cpuMinUsageDelta < 0 && server.ServerCpu > int32(bbb.Config.Resources.Cpu.MinCores) {
 		targetResource.Cpu.Direction = s.ScaleDown
-		targetResource.Cpu.Amount = int32(bbb.Config.Resources.Cpu.MinCores)
+		targetResource.Cpu.Reason = targetResource.Cpu.Reason + ",Rule 3"
+		cpuDec := cpuMinUsageDelta * float32(server.ServerCpu) / server.ServerCpuUsage
+		targetHeuristic := server.ServerCpu + int32(math.Floor(float64(cpuDec)))
+		targetResource.Cpu.Amount = int32(math.Max(float64(targetHeuristic), float64((bbb.Config.Resources.Cpu.MinCores))))
 	}
 
-	// Scaling memory
-	memMaxUsageDelta := server.ServerRamUsage - bbb.Config.Resources.Memory.MaxUsage
-
-	if server.ServerRam >= int32(bbb.Config.Resources.Memory.MinBytes) && memMaxUsageDelta <= 0 {
-		targetResource.Mem.Direction = s.ScaleNone
+	// Rule 4: Scale up memory if current ram is below configured minimum for the service
```

Comment:
I think this rule and the next should reuse rules 1 and 2, we don't need to differentiate by resource type